dev-packages/cli/src/download-plugins.ts

```diff
@@ -103,7 +109,12 @@ export default async function downloadPlugins(options: DownloadPluginsOptions =
             if (typeof pluginUrl !== 'string') {
                 continue;
             }
-            downloads.push(downloadPluginAsync(failures, plugin, pluginUrl, pluginsDir, packed));
+            const download = downloadPluginAsync(failures, plugin, pluginUrl, pluginsDir, packed);
+            if (parallel) {
+                downloads.push(download);
+            } else {
+                await download;
+            }
```

Comment:
This rewrite is a bit awkward in a context where we're still always declaring `downloads` and always `await`ing `Promise.all(downloads)`. I think it would be preferable to rewrite the implementation at a high level to respect the distinction between parallel and non-parallel downloads.

Reply:
I was wondering if we really want to do a full change just for this, as Promise.all([]) does return a already resolved promise so this is ... correct but looks a bit weird... 

I sort of don't like diverging a parallel vs serial download paths too much, as serial is just a special case of parallel :)

Do you have a better vision to make this code less clunky?

Reply:
My idea would be something like this:
```typescript
        const downloadPlugin = ([plugin, url]: [string, string]) => downloadPluginAsync(failures, plugin, url, pluginsDir, packed);
        const toDownload = Object.entries(pck.theiaPlugins).filter((entry: [string, unknown]): entry is [string, string] => typeof entry[1] === 'string');
        if (parallel) {
            await Promise.all(toDownload.map(downloadPlugin));
        } else {
            for (const entry of toDownload) {
                await downloadPlugin(entry);
            }
        }
```

However you'd like to do it, it will need to be done also for extension packs, since the code that handles those currently also does a parallel `Promise.all`.

Reply:
Yeah, I see the extension pack path now... got confused by the nested await but you are correct...

Reply:
I am not fluent with typescript , but given the need for the downloads to actually succeed, I would unconditionally chunk the downloads array into bins of size (say) 5 (or 1 if --no-parallel) and await one chunk after the other.

Reply:
... let's start with a simple yes/no and see how it turns out :) If you want to add chunking into it then feel free to try

Reply:
I'm not sure I see the benefit to this approach, since I don't think it's desirable (in general) to bail early. If I'm the developer, I'd like to get as many plugins as I can successfully get, and then address any failures. If we bail early, it's possible we'll discover additional failures the next time around that we could have tried to address after the first pass. That isn't exactly what the current code on `master` does, since it does stop at the first failure, but it will have initiated and actually execute other requests as well, so it comes close to what I think is the desired behavior.

Reply:
What I mean is: with chunking, the only effect of `--no-parallel` is on the chunk size. No other branches necessary.

Reply:
It's true no branching would be necessary, but all of the current code would have to be rewritten, since chunking requires serialization of a (variable number of) parallelized operations. If you have a concrete use case for chunks of sizes other than 1 (`--no-parallel` in this PR) or infinity (`--parallel` in this PR), please open an issue or contribute a PR to implement it.

Reply:
Rewrote the download parts into a couple of helper functions, and an interface type to carry the necessary plugin fields in a bit more structured form. 

This also makes the parallel/no-parallel decision in one function, if someone wants to add chunking later :)